File: ts/src/NexardaSDK.ts
// Nexarda Ts SDK

import { ConsoleEntity } from './entity/ConsoleEntity'
import { FranchisEntity } from './entity/FranchisEntity'
import { GameEntity } from './entity/GameEntity'
import { PlatformEntity } from './entity/PlatformEntity'
import { PriceEntity } from './entity/PriceEntity'
import { RetailerEntity } from './entity/RetailerEntity'
import { SearchEntity } from './entity/SearchEntity'
import { StudioEntity } from './entity/StudioEntity'
import { UserEntity } from './entity/UserEntity'
import { WidgetEntity } from './entity/WidgetEntity'

export type * from './NexardaTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { NexardaEntityBase } from './NexardaEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class NexardaSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _console?: ConsoleEntity

  // Idiomatic facade: `client.console.list()` / `client.console.load({ id })`.
  get console(): ConsoleEntity {
    return (this._console ??= new ConsoleEntity(this, undefined))
  }

  /** @deprecated Use `client.console` instead. */
  Console(data?: any) {
    const self = this
    return new ConsoleEntity(self,data)
  }


  _franchis?: FranchisEntity

  // Idiomatic facade: `client.franchis.list()` / `client.franchis.load({ id })`.
  get franchis(): FranchisEntity {
    return (this._franchis ??= new FranchisEntity(this, undefined))
  }

  /** @deprecated Use `client.franchis` instead. */
  Franchis(data?: any) {
    const self = this
    return new FranchisEntity(self,data)
  }


  _game?: GameEntity

  // Idiomatic facade: `client.game.list()` / `client.game.load({ id })`.
  get game(): GameEntity {
    return (this._game ??= new GameEntity(this, undefined))
  }

  /** @deprecated Use `client.game` instead. */
  Game(data?: any) {
    const self = this
    return new GameEntity(self,data)
  }


  _platform?: PlatformEntity

  // Idiomatic facade: `client.platform.list()` / `client.platform.load({ id })`.
  get platform(): PlatformEntity {
    return (this._platform ??= new PlatformEntity(this, undefined))
  }

  /** @deprecated Use `client.platform` instead. */
  Platform(data?: any) {
    const self = this
    return new PlatformEntity(self,data)
  }


  _price?: PriceEntity

  // Idiomatic facade: `client.price.list()` / `client.price.load({ id })`.
  get price(): PriceEntity {
    return (this._price ??= new PriceEntity(this, undefined))
  }

  /** @deprecated Use `client.price` instead. */
  Price(data?: any) {
    const self = this
    return new PriceEntity(self,data)
  }


  _retailer?: RetailerEntity

  // Idiomatic facade: `client.retailer.list()` / `client.retailer.load({ id })`.
  get retailer(): RetailerEntity {
    return (this._retailer ??= new RetailerEntity(this, undefined))
  }

  /** @deprecated Use `client.retailer` instead. */
  Retailer(data?: any) {
    const self = this
    return new RetailerEntity(self,data)
  }


  _search?: SearchEntity

  // Idiomatic facade: `client.search.list()` / `client.search.load({ id })`.
  get search(): SearchEntity {
    return (this._search ??= new SearchEntity(this, undefined))
  }

  /** @deprecated Use `client.search` instead. */
  Search(data?: any) {
    const self = this
    return new SearchEntity(self,data)
  }


  _studio?: StudioEntity

  // Idiomatic facade: `client.studio.list()` / `client.studio.load({ id })`.
  get studio(): StudioEntity {
    return (this._studio ??= new StudioEntity(this, undefined))
  }

  /** @deprecated Use `client.studio` instead. */
  Studio(data?: any) {
    const self = this
    return new StudioEntity(self,data)
  }


  _user?: UserEntity

  // Idiomatic facade: `client.user.list()` / `client.user.load({ id })`.
  get user(): UserEntity {
    return (this._user ??= new UserEntity(this, undefined))
  }

  /** @deprecated Use `client.user` instead. */
  User(data?: any) {
    const self = this
    return new UserEntity(self,data)
  }


  _widget?: WidgetEntity

  // Idiomatic facade: `client.widget.list()` / `client.widget.load({ id })`.
  get widget(): WidgetEntity {
    return (this._widget ??= new WidgetEntity(this, undefined))
  }

  /** @deprecated Use `client.widget` instead. */
  Widget(data?: any) {
    const self = this
    return new WidgetEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new NexardaSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return NexardaSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'Nexarda' }
  }

  toString() {
    return 'Nexarda ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = NexardaSDK


export {
  stdutil,

  BaseFeature,
  NexardaEntityBase,

  NexardaSDK,
  SDK,
}

File: go/core/nexarda_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/nexarda-sdk/go/utility/struct"
)

type NexardaSDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewNexardaSDK(options map[string]any) *NexardaSDK {
	sdk := &NexardaSDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features in the resolved order (MakeOptions puts an explicit array
	// order first, else defaults to test-first). Ordering matters: the `test`
	// feature installs the base mock transport and the transport features
	// (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
	// must be added before them to sit at the base of the chain.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		if fo, ok := vs.GetPath([]any{"__derived__", "featureorder"}, sdk.options).([]any); ok {
			for _, n := range fo {
				fname, _ := n.(string)
				fopts := ToMapAny(featureOpts[fname])
				if fopts != nil {
					if active, ok := fopts["active"]; ok {
						if ab, ok := active.(bool); ok && ab {
							sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
						}
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *NexardaSDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *NexardaSDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *NexardaSDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *NexardaSDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *NexardaSDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// Console returns a Console entity bound to this client.
// Idiomatic usage: client.Console(nil).List(nil, nil) or
// client.Console(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Console(data map[string]any) NexardaEntity {
	return NewConsoleEntityFunc(sdk, data)
}


// Franchis returns a Franchis entity bound to this client.
// Idiomatic usage: client.Franchis(nil).List(nil, nil) or
// client.Franchis(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Franchis(data map[string]any) NexardaEntity {
	return NewFranchisEntityFunc(sdk, data)
}


// Game returns a Game entity bound to this client.
// Idiomatic usage: client.Game(nil).List(nil, nil) or
// client.Game(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Game(data map[string]any) NexardaEntity {
	return NewGameEntityFunc(sdk, data)
}


// Platform returns a Platform entity bound to this client.
// Idiomatic usage: client.Platform(nil).List(nil, nil) or
// client.Platform(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Platform(data map[string]any) NexardaEntity {
	return NewPlatformEntityFunc(sdk, data)
}


// Price returns a Price entity bound to this client.
// Idiomatic usage: client.Price(nil).List(nil, nil) or
// client.Price(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Price(data map[string]any) NexardaEntity {
	return NewPriceEntityFunc(sdk, data)
}


// Retailer returns a Retailer entity bound to this client.
// Idiomatic usage: client.Retailer(nil).List(nil, nil) or
// client.Retailer(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Retailer(data map[string]any) NexardaEntity {
	return NewRetailerEntityFunc(sdk, data)
}


// Search returns a Search entity bound to this client.
// Idiomatic usage: client.Search(nil).List(nil, nil) or
// client.Search(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Search(data map[string]any) NexardaEntity {
	return NewSearchEntityFunc(sdk, data)
}


// Studio returns a Studio entity bound to this client.
// Idiomatic usage: client.Studio(nil).List(nil, nil) or
// client.Studio(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Studio(data map[string]any) NexardaEntity {
	return NewStudioEntityFunc(sdk, data)
}


// User returns a User entity bound to this client.
// Idiomatic usage: client.User(nil).List(nil, nil) or
// client.User(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) User(data map[string]any) NexardaEntity {
	return NewUserEntityFunc(sdk, data)
}


// Widget returns a Widget entity bound to this client.
// Idiomatic usage: client.Widget(nil).List(nil, nil) or
// client.Widget(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *NexardaSDK) Widget(data map[string]any) NexardaEntity {
	return NewWidgetEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *NexardaSDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewNexardaSDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}
